File: src/SerializerLib/SerializerLib/Tests/SerializerTest.cs
using System;
using System.Collections.Generic;

namespace SerializerLib
{
    namespace MAP
    {
        using NetworkObject = List<INetworkType>;

        public class SerializerTest
        {
            public SerializerTest()
            {
            }

            public bool ByteTypeTest()
            {
                NetworkObject sequence = new NetworkObject(){
                new NetCommand(0x01, 102),
                new NetByte(42, "answer"),
                new NetByte(128, "halfbyte"),
                new NetByte(255, "fullbyte")};

                var binaryData = BinaryUtils.Encode(sequence);
                var deserialized = BinaryUtils.DecodeAsMap(binaryData.ToArray(), binaryData.Count - 1);

                var half = ((NetByte)deserialized["halfbyte"]).GetValue();
                var full = ((NetByte)deserialized["fullbyte"]).GetValue();
                var answer = ((NetByte)deserialized["answer"]).GetValue();


                return (half == 128 && full == 255 && answer == 42);
            }

            public bool DynamicTypeArrayTest()
            {
                NetworkObject sequence = new NetworkObject(){
                new NetByte(42, "byteVal"),
                new NetString("some string lol", "testStr"),
                new NetFloat(0.33f, "floating")};

                var arrayObj = new NetArray(sequence, "arrayTest");
                var serializedArrayVector = arrayObj.Serialize();
                var objMap = BinaryUtils.DecodeAsMap(serializedArrayVector.ToArray(), serializedArrayVector.Count - 1);

                var arr = ((NetArray)objMap["arrayTest"]).GetValues();
                var testbyte = ((NetByte)arr[0]).GetValue();
                var str = ((NetString)arr[1]).GetValue();
                var flt = ((NetFloat)arr[2]).GetValue();

                return testbyte == 42;
            }

            public bool StringTypeTest()
            {
                NetworkObject sequence = new NetworkObject(){
                new NetString("Hello world", "hello"),
                new NetString("This is a large text, can contain 255 characters", "longText")
                };
                var serializedStrings = BinaryUtils.Encode(sequence);
                var deserialized = BinaryUtils.DecodeAsMap(serializedStrings.ToArray(), serializedStrings.Count - 1);

                var hello = ((NetString)deserialized["hello"]).GetValue();
                var longText = ((NetString)deserialized["longText"]).GetValue();

                return hello.CompareTo("Hello world") == 0;
            }

            public bool FloatTypeTest()
            {
                NetworkObject sequence = new NetworkObject(){
                new NetFloat(0.006f, "tiny6"),
                new NetFloat(42.0f, "abc"),
                new NetFloat(128.0f, "half"),
                new NetFloat(255.0f, "full")};

                var serializedFloats = BinaryUtils.Encode(sequence);
                var deserialized = BinaryUtils.DecodeAsMap(serializedFloats.ToArray(), serializedFloats.Count - 1);

                var tinyVal = ((NetFloat)deserialized["tiny6"]).GetValue();
                var halfVal = ((NetFloat)deserialized["half"]).GetValue();

                return tinyVal == 0.006f && halfVal == 128.0f;
            }

            public bool IntTypeTest()
            {
                NetworkObject sequence = new NetworkObject(){
                new NetInt(777, "luck"),
                new NetInt(666, "hell"),
                new NetInt(12345678, "large")};

                var serializedInts = BinaryUtils.Encode(sequence);
                var deserialized = BinaryUtils.DecodeAsMap(serializedInts.ToArray(), serializedInts.Count - 1);
                var luck = ((NetInt)deserialized["luck"]).GetValue();
                var hell = ((NetInt)deserialized["hell"]).GetValue();
                var large = ((NetInt)deserialized["large"]).GetValue();

                return luck == 777 && hell == 666 && large == 12345678;
            }

            public bool ComplexArrayObjectTest()
            {
                NetworkObject poolData = new NetworkObject(){
                new NetInt(666, "PoolId"),
                new NetString("TESTING_POOL", "PoolName")};

                NetworkObject commandPayload = new NetworkObject(){
                new NetArray(poolData, "pool"),
                new NetArray(poolData, "pool2"),
                new NetArray(poolData, "pool3")
                };

                var serializedObject = BinaryUtils.Encode(commandPayload);
                var deserializedObjectMap = BinaryUtils.DecodeAsMap(serializedObject.ToArray(), serializedObject.Count - 1);


                var poolDataArr = ((NetArray)deserializedObjectMap["pool2"]).GetValues();
                var poolId = ((NetInt)poolDataArr[0]).GetValue();
                var poolName = ((NetString)poolDataArr[1]).GetValue();

                return poolId == 666 && poolName.CompareTo("TESTING_POOL") == 0;
            }

            public bool StaticTypeArrayTest()
            {
                // FLOAT TEST TYPE,first type cos i like floats
                NetworkObject sequence = new NetworkObject(){
                    new NetFloat(0.42f, ""),
                    new NetFloat(1.42f, ""),
                    new NetFloat(2.42f, "")
                    };

                var arrayObj = new NetStaticArray(sequence, NetworkType.FLOAT, "vector3f");
                var serializedArrayVector = arrayObj.Serialize();
                var objMap = BinaryUtils.DecodeAsMap(serializedArrayVector.ToArray(), serializedArrayVector.Count - 1);

                var arr = ((NetStaticArray)objMap["vector3f"]).GetValues();
                var f1 = ((NetFloat)arr[0]).GetValue();
                var f2 = ((NetFloat)arr[1]).GetValue();
                var f3 = ((NetFloat)arr[2]).GetValue();

                //STRING TYPE ARRAY TEST
                NetworkObject sequence2 = new NetworkObject(){
                    new NetString("AAAA", ""),
                    new NetString("BBBBbBBB", ""),
                    new NetString("CCCCcccccc", "")
                    };

                var arrayObj2 = new NetStaticArray(sequence2, NetworkType.STRING, "stringVector");
                var serializedArrayVector2 = arrayObj2.Serialize();
                var objMap2 = BinaryUtils.DecodeAsMap(serializedArrayVector2.ToArray(), serializedArrayVector2.Count - 1);

                var arr2 = ((NetStaticArray)objMap2["stringVector"]).GetValues();
                var s1 = ((NetString)arr2[0]).GetValue();
                var s2 = ((NetString)arr2[1]).GetValue();
                var s3 = ((NetString)arr2[2]).GetValue();

                //INT TYPE ARRAY TEST
                NetworkObject sequence3 = new NetworkObject(){
                    new NetInt(42, ""),
                    new NetInt(41, ""),
                    new NetInt(12345678, "")
                    };

                var arrayObj3 = new NetStaticArray(sequence3, NetworkType.INT, "intVector");
                var serializedArrayVector3 = arrayObj3.Serialize();
                var objMap3 = BinaryUtils.DecodeAsMap(serializedArrayVector3.ToArray(), serializedArrayVector3.Count - 1);

                var arr3 = ((NetStaticArray)objMap3["intVector"]).GetValues();
                var i1 = ((NetInt)arr3[0]).GetValue();
                var i2 = ((NetInt)arr3[1]).GetValue();
                var i3 = ((NetInt)arr3[2]).GetValue();

                //BYTE TYPE ARRAY TEST
                NetworkObject sequence4 = new NetworkObject(){
                    new NetByte(42, ""),
                    new NetByte(41, ""),
                    new NetByte(40, "")
                    };

                var arrayObj4 = new NetStaticArray(sequence4, NetworkType.BYTE, "byteVector");
                var serializedArrayVector4 = arrayObj4.Serialize();
                var objMap4 = BinaryUtils.DecodeAsMap(serializedArrayVector4.ToArray(), serializedArrayVector4.Count - 1);

                var arr4 = ((NetStaticArray)objMap4["byteVector"]).GetValues();
                var b1 = ((NetByte)arr4[0]).GetValue();
                var b2 = ((NetByte)arr4[1]).GetValue();
                var b3 = ((NetByte)arr4[2]).GetValue();

                //Dynamic type array ( MOST COMPLEX OBJECT UNTIL NOW... :> )
                NetworkObject somePair = new NetworkObject(){
                    new NetInt(123452, "id"),
                    new NetString("testing", "content")
                };

                NetworkObject pairArray = new NetworkObject(){
                    new NetArray(somePair, ""),
                    new NetArray(somePair, ""),
                    new NetArray(somePair, "")
                };

                var arrayObj5 = new NetStaticArray(pairArray, NetworkType.ARRAY, "dynamicArrayVector");
                var serializedArrayVector5 = arrayObj5.Serialize();
                var objMap5 = BinaryUtils.DecodeAsMap(serializedArrayVector5.ToArray(), serializedArrayVector5.Count - 1);

                var arr5 = ((NetStaticArray)objMap5["dynamicArrayVector"]).GetValues();
                var a1 = ((NetArray)arr5[0]).GetValues();
                var a2 = ((NetArray)arr5[1]).GetValues();
                var a3 = ((NetArray)arr5[2]).GetValues();

                return true;
            }

            public bool Check()
            {
                bool allRight = false;
                // allRight = ByteTypeTest();
                // allRight = DynamicTypeArrayTest();
                // allRight = StringTypeTest();
                // allRight = FloatTypeTest();
                // allRight = IntTypeTest();
                // allRight = ComplexArrayObjectTest();
                allRight = StaticTypeArrayTest();
                return allRight;
            }
        }

    }
}

File: src/Tests/Serializer/SerializerTest.cpp
#ifndef SERIALIZER_TEST
#define SERIALIZER_TEST

#include <BinaryUtils.hpp>
#include <gtest/gtest.h>

#include <memory>
#include <vector>
using namespace MAP;

TEST(MapSerializer, ByteTypeTest)
{
    NetworkObject sequence = {
        std::make_shared<MAP::NetCommand>(1, 102),
        std::make_shared<MAP::NetByte>(42, "answer"),
        std::make_shared<MAP::NetByte>(128, "halfbyte"),
        std::make_shared<MAP::NetByte>(255, "fullbyte")};

    auto binaryData = BinaryUtils::Encode(sequence);
    auto deserialized = BinaryUtils::DecodeAsMap(binaryData.data(), binaryData.size());

    auto half = std::dynamic_pointer_cast<MAP::NetByte>(deserialized["halfbyte"]);
    auto full = std::dynamic_pointer_cast<MAP::NetByte>(deserialized["fullbyte"]);
    auto answer = std::dynamic_pointer_cast<MAP::NetByte>(deserialized["answer"]);

    auto test1 = half->GetValue();
    auto test2 = full->GetValue();
    auto test3 = answer->GetValue();

    EXPECT_EQ(test1 == 128 && test2 == 255 && test3 == 42, true);
}

TEST(MapSerializer, DynamicArrayTypeTest)
{
    NetworkObject sequence = {
        std::make_shared<MAP::NetByte>(42, "byteVal"),
        std::make_shared<MAP::NetFloat>(0.33f, "decimal"),
        std::make_shared<MAP::NetString>("XXXXXXXXXXXXXXXXXXXXXXXXXX", "someString")};

    auto arrayObj = std::make_shared<MAP::NetArray>(sequence, "arrayTest");
    auto serializedArrayVector = arrayObj->Serialize();
    auto objMap = BinaryUtils::DecodeAsMap(serializedArrayVector.data(), serializedArrayVector.size());
    auto arr = std::dynamic_pointer_cast<MAP::NetArray>(objMap["arrayTest"])->GetValues();

    auto byte = BinaryUtils::Get<MAP::NetByte>(arr, "byteVal")->GetValue();
    auto floatVal = BinaryUtils::Get<MAP::NetFloat>(arr, "decimal")->GetValue();
    auto stringVal = BinaryUtils::Get<MAP::NetString>(arr, "someString")->GetValue();

    EXPECT_EQ(byte, 42);
    EXPECT_EQ(floatVal, 0.33f);
    EXPECT_STREQ(stringVal.c_str(), "XXXXXXXXXXXXXXXXXXXXXXXXXX");
}

TEST(MapSerializer, StringTypeTest)
{
    NetworkObject sequence = {
        std::make_shared<MAP::NetString>("Hello world", "hello"),
        std::make_shared<MAP::NetString>("This is a large text, can contain 255 characters", "longText")};
    auto serializedStrings = BinaryUtils::Encode(sequence);
    auto deserialized = BinaryUtils::DecodeAsMap(serializedStrings.data(), serializedStrings.size());
    auto hello = std::dynamic_pointer_cast<MAP::NetString>(deserialized["hello"]);
    auto longText = std::dynamic_pointer_cast<MAP::NetString>(deserialized["longText"]);
    auto helloString = hello->GetValue();
    auto largeString = longText->GetValue();
    EXPECT_TRUE(helloString.compare("Hello world") == 0);
    EXPECT_TRUE(largeString.compare("This is a large text, can contain 255 characters") == 0);
}

TEST(MapSerializer, FloatTypeTest)
{
    NetworkObject sequence = {
        std::make_shared<MAP::NetFloat>(0.006f, "tiny6"),
        std::make_shared<MAP::NetFloat>(42.0f, "abc"),
        std::make_shared<MAP::NetFloat>(128.0f, "half"),
        std::make_shared<MAP::NetFloat>(255.0f, "full")};

    auto serializedFloats = BinaryUtils::Encode(sequence);
    auto deserialized = BinaryUtils::DecodeAsMap(serializedFloats.data(), serializedFloats.size());

    auto tinyVal = BinaryUtils::Get<MAP::NetFloat>(deserialized, "tiny6");
    auto halfVal = BinaryUtils::Get<MAP::NetFloat>(deserialized, "half");

    float val = tinyVal->GetValue();
    float val2 = halfVal->GetValue();

    EXPECT_EQ(val, 0.006f);
    EXPECT_EQ(val2, 128.0f);
}

TEST(MapSerializer, Int32TypeTest)
{
    NetworkObject sequence = {
        std::make_shared<MAP::NetInt>(777, "luck"),
        std::make_shared<MAP::NetInt>(666, "hell"),
        std::make_shared<MAP::NetInt>(12345678, "large")};

    auto serializedInts = BinaryUtils::Encode(sequence);
    auto deserialized = BinaryUtils::DecodeAsMap(serializedInts.data(), serializedInts.size());

    auto luck = BinaryUtils::Get<MAP::NetInt>(deserialized, "luck");
    auto hell = BinaryUtils::Get<MAP::NetInt>(deserialized, "hell");
    auto large = BinaryUtils::Get<MAP::NetInt>(deserialized, "large");

    int val = luck->GetValue();
    int val2 = hell->GetValue();
    int val3 = large->GetValue();

    EXPECT_EQ(val, 777);
    EXPECT_EQ(val2, 666);
    EXPECT_EQ(val3, 12345678);
}

TEST(MapSerializer, ComplexArrayObjectTest)
{
    NetworkObject poolData{
        std::make_shared<MAP::NetInt>(666, "poolId"),
        std::make_shared<MAP::NetString>("Example name", "poolName")};

    NetworkObject
        commandPayload{
            std::make_shared<MAP::NetArray>(poolData, "AAAAA"),
            std::make_shared<MAP::NetArray>(poolData, "BBBB")};

    auto serializedObject = BinaryUtils::Encode(commandPayload);
    auto deserializedObjectMap = BinaryUtils::DecodeAsMap(serializedObject.data(), serializedObject.size());
    //Get pool obj3
    auto poolDataArr = BinaryUtils::Get<MAP::NetArray>(deserializedObjectMap, "BBBB")->GetValues();
    auto poolId = BinaryUtils::Get<MAP::NetInt>(poolDataArr, "poolId")->GetValue();
    auto poolName = BinaryUtils::Get<MAP::NetString>(poolDataArr, "poolName")->GetValue();
    EXPECT_EQ(poolId, 666);
    EXPECT_TRUE(poolName.compare("Example name") == 0);
}

TEST(MapSerializer, StaticArrayTest)
{
    //Float testing (emulates a vector 3)
    NetworkObject sequence = {
        std::make_shared<MAP::NetFloat>(0.42f, ""),
        std::make_shared<MAP::NetFloat>(0.33f, ""),
        std::make_shared<MAP::NetFloat>(0.42069f, "")};

    auto arrayObj = std::make_shared<MAP::NetStaticArray>(sequence, NetworkType::FLOAT, "sarrayTest");
    auto serializedArrayVector = arrayObj->Serialize();
    auto objMap = BinaryUtils::DecodeAsMap(serializedArrayVector.data(), serializedArrayVector.size());
    auto arr = std::dynamic_pointer_cast<MAP::NetStaticArray>(objMap["sarrayTest"])->GetValues();

    auto first = std::dynamic_pointer_cast<MAP::NetFloat>(arr.at(0))->GetValue();
    auto second = std::dynamic_pointer_cast<MAP::NetFloat>(arr.at(1))->GetValue();
    auto third = std::dynamic_pointer_cast<MAP::NetFloat>(arr.at(2))->GetValue();

    EXPECT_EQ(first, 0.42f);
    EXPECT_EQ(second, 0.33f);
    EXPECT_EQ(third, 0.42069f);

    //String type test
    NetworkObject sequence2 = {
        std::make_shared<MAP::NetString>("HELLO", ""),
        std::make_shared<MAP::NetString>("world", ""),
        std::make_shared<MAP::NetString>("XxXx  XxXxXx", "")};

    auto arrayObj2 = std::make_shared<MAP::NetStaticArray>(sequence2, NetworkType::STRING, "sarrayTest2");
    auto serializedArrayVector2 = arrayObj2->Serialize();
    auto objMap2 = BinaryUtils::DecodeAsMap(serializedArrayVector2.data(), serializedArrayVector2.size());
    auto arr2 = std::dynamic_pointer_cast<MAP::NetStaticArray>(objMap2["sarrayTest2"])->GetValues();

    auto firstString = std::dynamic_pointer_cast<MAP::NetString>(arr2.at(0))->GetValue();
    auto secondString = std::dynamic_pointer_cast<MAP::NetString>(arr2.at(1))->GetValue();
    auto thirdString = std::dynamic_pointer_cast<MAP::NetString>(arr2.at(2))->GetValue();

    EXPECT_TRUE(firstString.compare("HELLO") == 0);
    EXPECT_TRUE(secondString.compare("world") == 0);
    EXPECT_TRUE(thirdString.compare("XxXx  XxXxXx") == 0);

    //Int type test
    NetworkObject sequence3 = {
        std::make_shared<MAP::NetInt>(111111, ""),
        std::make_shared<MAP::NetInt>(222222, ""),
        std::make_shared<MAP::NetInt>(333333, "")};

    auto arrayObj3 = std::make_shared<MAP::NetStaticArray>(sequence3, NetworkType::INT, "sarrayTest3");
    auto serializedArrayVector3 = arrayObj3->Serialize();
    auto objMap3 = BinaryUtils::DecodeAsMap(serializedArrayVector3.data(), serializedArrayVector3.size());
    auto arr3 = std::dynamic_pointer_cast<MAP::NetStaticArray>(objMap3["sarrayTest3"])->GetValues();

    auto firstInt = std::dynamic_pointer_cast<MAP::NetInt>(arr3.at(0))->GetValue();
    auto secondInt = std::dynamic_pointer_cast<MAP::NetInt>(arr3.at(1))->GetValue();
    auto thirdInt = std::dynamic_pointer_cast<MAP::NetInt>(arr3.at(2))->GetValue();
    EXPECT_EQ(firstInt, 111111);
    EXPECT_EQ(secondInt, 222222);
    EXPECT_EQ(thirdInt, 333333);

    //Byte type test
    NetworkObject sequence4 = {
        std::make_shared<MAP::NetByte>(1, ""),
        std::make_shared<MAP::NetByte>(128, ""),
        std::make_shared<MAP::NetByte>(255, "")};

    auto arrayObj4 = std::make_shared<MAP::NetStaticArray>(sequence4, NetworkType::BYTE, "sarrayTest4");
    auto serializedArrayVector4 = arrayObj4->Serialize();
    auto objMap4 = BinaryUtils::DecodeAsMap(serializedArrayVector4.data(), serializedArrayVector4.size());
    auto arr4 = std::dynamic_pointer_cast<MAP::NetStaticArray>(objMap4["sarrayTest4"])->GetValues();

    auto firstByte = std::dynamic_pointer_cast<MAP::NetByte>(arr4.at(0))->GetValue();
    auto secondByte = std::dynamic_pointer_cast<MAP::NetByte>(arr4.at(1))->GetValue();
    auto thirdByte = std::dynamic_pointer_cast<MAP::NetByte>(arr4.at(2))->GetValue();

    EXPECT_EQ(firstByte, 1);
    EXPECT_EQ(secondByte, 128);
    EXPECT_EQ(thirdByte, 255);
}

TEST(MapSerializer, TypeAssignmentTest)
{
    // THIS IS A ASSIGNMENT TEST WITHOUT OPERATOR OVERLOAD IN THE INTERFACE(TODO!!!)
    std::map<std::string, std::shared_ptr<MAP::INetworkType>> testMap{
        std::make_pair("element1", std::make_shared<MAP::NetInt>(666, "PoolId")),
        std::make_pair("element2", std::make_shared<MAP::NetInt>(666, "PoolId"))};

    auto someExternalNetworkType = std::make_shared<MAP::NetInt>(33333, "PoolId");
    testMap["element1"] = someExternalNetworkType;
    int testMapVal = std::dynamic_pointer_cast<MAP::NetInt>(testMap["element1"])->GetValue();
    EXPECT_EQ(testMapVal, 33333);
}

#endif